src/main/java/org/jenkinsci/plugins/workflow/libs/SCMSourceRetriever.java

```diff
@@ -109,7 +111,32 @@ static void doRetrieve(String name, boolean changelog, @Nonnull SCM scm, FilePat
             throw new IOException(node.getDisplayName() + " may be offline");
         }
         try (WorkspaceList.Lease lease = computer.getWorkspaceList().allocate(dir)) {
-            delegate.checkout(run, lease.path, listener, node.createLauncher(listener));
+            for (int retryCount = Jenkins.getInstance().getScmCheckoutRetryCount(); retryCount >= 0; retryCount--) {
+                try {
+                    delegate.checkout(run, lease.path, listener, node.createLauncher(listener));
+                    break;
+                }
+                catch (AbortException e) {
+                    // abort exception might have a null message.
+                    // If so, just skip echoing it.
+                    if (e.getMessage() != null) {
+                        listener.error(e.getMessage());
+                    }
+                }
+                catch (InterruptedIOException e) {
+                    throw e;
+                }
+                catch (IOException e) {
+                    // checkout error not yet reported
+                    listener.error("Checkout failed").println(Functions.printThrowable(e).trim()); // TODO 2.43+ use Functions.printStackTrace
+                }
+
+                if (retryCount == 0)   // all attempts failed
+                    throw new AbortException("Maximum checkout retry attempts reached, aborting");
+
+                listener.getLogger().println("Retrying after 10 seconds");
+                Thread.sleep(10000);
```

Comment:
Will this sleep cause problems if it runs as part of `LibraryStep`? We extend `AbstractSynchronousNonBlockingStepExecution`, so I guess it won't block the main CPS VM, but would be great for one of the more experienced devs to confirm.

Reply:
Might be worth going with a `Timer` like in https://github.com/jenkinsci/workflow-basic-steps-plugin/blob/906497e661b67c7a18a65c43ffd526f4dd7a5938/src/main/java/org/jenkinsci/plugins/workflow/steps/SleepStep.java#L99